CI complains of outdated lockfile but says the lockfile is up to date

The CLI says the lockfile is up to date but it still failing when it runs on CI. Running `pnpm i` locally doesn't create any updates to the lockfile.

### pnpm version: 7.8.0

The issue exists since the release of 7.6.0. It works fine in 7.5.2

### Expected behavior:

CI doesn't complain about outdated file when the lockfile is updated.

### Actual behavior:

CI complains the lockfile is outdated

```
pnpm install --prefer-offline

Scope: all 21 workspace projects
Lockfile is up-to-date, resolution step is skipped
 ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up-to-date with packages/eslint-plugin/package.json

Note that in CI environments this setting is true by default. If you still need to run install in such cases, use "pnpm install --no-frozen-lockfile"

Exited with code exit status 1
```

### Additional information:

 - `node -v` prints: 16.16.0
 - Windows, macOS, or Linux?: OSX/Linux

@zkochan removing lockfile and node_modules did not work. What helps is moving a peer dependency from `peerDependencies` to `devDependencies`. For example I had following error

```
ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up to date with config/stylelint-config/package.json
```

and `stylelint-config/package.json` had `autoprefixer` in `peerDependencies` and once I moved it to `devDependencies` the error disappeared. But I have more cases like this...

--- a/.changeset/twenty-dragons-explode.md
+++ b/.changeset/twenty-dragons-explode.md
@@ -1,0 +1,7 @@
+---
+"@pnpm/core": patch
+"@pnpm/resolve-dependencies": patch
+"pnpm": patch
+---
+
+Auto installing a peer dependency in a workspace that also has it as a dev dependency in another project [#5144](https://github.com/pnpm/pnpm/issues/5144).

--- a/packages/core/test/install/autoInstallPeers.ts
+++ b/packages/core/test/install/autoInstallPeers.ts
@@ -1,3 +1,5 @@
+import path from 'path'
+import assertProject from '@pnpm/assert-project'
 import { addDependenciesToPackage, install, mutateModules } from '@pnpm/core'
 import { prepareEmpty, preparePackages } from '@pnpm/prepare'
 import { addDistTag, REGISTRY_MOCK_PORT } from '@pnpm/registry-mock'
@@ -225,4 +227,42 @@ test('automatically install root peer dependencies', async () => {
       'is-negative': '1.0.1',
     })
   }
+})
+
+test('automatically install peer dependency when it is a dev dependency in another workspace project', async () => {
+  prepareEmpty()
+
+  await mutateModules([
+    {
+      buildIndex: 0,
+      manifest: {
+        name: 'project-1',
+        devDependencies: {
+          'is-positive': '1.0.0',
+        },
+      },
+      mutation: 'install',
+      rootDir: path.resolve('project-1'),
+    },
+    {
+      buildIndex: 0,
+      manifest: {
+        name: 'project-2',
+        peerDependencies: {
+          'is-positive': '1.0.0',
+        },
+      },
+      mutation: 'install',
+      rootDir: path.resolve('project-2'),
+    },
+  ], await testDefaults({ autoInstallPeers: true }))
+
+  const project = assertProject(process.cwd())
+  const lockfile = await project.readLockfile()
+  expect(lockfile.importers['project-1'].devDependencies).toStrictEqual({
+    'is-positive': '1.0.0',
+  })
+  expect(lockfile.importers['project-2'].dependencies).toStrictEqual({
+    'is-positive': '1.0.0',
+  })
 })

--- a/packages/resolve-dependencies/src/index.ts
+++ b/packages/resolve-dependencies/src/index.ts
@@ -135,9 +135,11 @@ export default async function (
     }
     if (opts.autoInstallPeers) {
       if (updatedManifest?.peerDependencies) {
-        updatedManifest.dependencies = {
-          ...updatedManifest.peerDependencies,
-          ...updatedManifest.dependencies,
+        const allDeps = getAllDependenciesFromManifest(updatedManifest)
+        for (const [peerName, peerRange] of Object.entries(updatedManifest.peerDependencies)) {
+          if (allDeps[peerName]) continue
+          updatedManifest.dependencies ??= {}
+          updatedManifest.dependencies[peerName] = peerRange
         }
       }
     }

--- a/packages/resolve-dependencies/src/index.ts
+++ b/packages/resolve-dependencies/src/index.ts
@@ -133,6 +133,14 @@ export default async function (
         updated: project.manifest,
       })
     }
+    if (opts.autoInstallPeers) {
+      if (updatedManifest?.peerDependencies) {
+        updatedManifest.dependencies = {
+          ...updatedManifest.peerDependencies,
+          ...updatedManifest.dependencies,
+        }
+      }
+    }
 
     if (updatedManifest != null) {
       const projectSnapshot = opts.wantedLockfile.importers[project.id]

--- a/packages/resolve-dependencies/src/index.ts
+++ b/packages/resolve-dependencies/src/index.ts
@@ -133,18 +133,18 @@ export default async function (
         updated: project.manifest,
       })
     }
-    if (opts.autoInstallPeers) {
-      if (updatedManifest?.peerDependencies) {
-        const allDeps = getAllDependenciesFromManifest(updatedManifest)
-        for (const [peerName, peerRange] of Object.entries(updatedManifest.peerDependencies)) {
-          if (allDeps[peerName]) continue
-          updatedManifest.dependencies ??= {}
-          updatedManifest.dependencies[peerName] = peerRange
-        }
-      }
-    }
 
     if (updatedManifest != null) {
+      if (opts.autoInstallPeers) {
+        if (updatedManifest.peerDependencies) {
+          const allDeps = getAllDependenciesFromManifest(updatedManifest)
+          for (const [peerName, peerRange] of Object.entries(updatedManifest.peerDependencies)) {
+            if (allDeps[peerName]) continue
+            updatedManifest.dependencies ??= {}
+            updatedManifest.dependencies[peerName] = peerRange
+          }
+        }
+      }
       const projectSnapshot = opts.wantedLockfile.importers[project.id]
       opts.wantedLockfile.importers[project.id] = addDirectDependenciesToLockfile(
         updatedManifest,
@@ -344,14 +344,7 @@ function addDirectDependenciesToLockfile (
     return acc
   }, {})
 
-  let allDepsObj = getAllDependenciesFromManifest(newManifest)
-  if (autoInstallPeers) {
-    allDepsObj = {
-      ...newManifest.peerDependencies,
-      ...allDepsObj,
-    }
-  }
-  const allDeps = Array.from(new Set(Object.keys(allDepsObj)))
+  const allDeps = Array.from(new Set(Object.keys(getAllDependenciesFromManifest(newManifest))))
 
   for (const alias of allDeps) {
     if (directDependenciesByAlias[alias]) {

